manager/controllers/app/moduleinstance.go

```diff
@@ -104,41 +110,65 @@ func (m *ModuleManager) SelectModuleInstances(item modules.DataInfo) ([]modules.
 	// DataStore for destination will be determined if an implicit copy is required
 	sinkDataStore = nil
 
-	// select a read module that supports user interface requirements
-	// actions are not checked since they are not necessarily done by the read module
-	readSelector = &modules.Selector{Flow: app.Read,
-		Destination:  sink,
-		Actions:      make([]pb.EnforcementAction, 0),
-		Source:       nil,
-		Dependencies: make([]*app.M4DModule, 0),
-		Module:       nil,
-		Message:      ""}
-	if !readSelector.SelectModule(m.Modules) {
-		m.Log.Info(item.AssetID + " : " + readSelector.GetError())
-		return instances, errors.New(readSelector.GetError())
+	var readSelector, copySelector *modules.Selector
+	actionsOnRead, readRequired := item.Actions[pb.AccessOperation_READ]
+	var interfaces []*app.InterfaceDetails
+	var copyRequired bool
+	var actionsOnCopy []pb.EnforcementAction
+	if readRequired {
+		// READ FLOW
+		m.Log.Info("Select read path for " + item.AssetID)
+		// Select a module that supports READ flow, supports actions-on-read, has the required dependency modules (recursively), with API = sink.
+		if !actionsOnRead.Allowed {
+			return instances, errors.New(actionsOnRead.Message)
+		}
+		m.Log.Info("Finding modules for " + item.AssetID)
+
+		// select a read module that supports user interface requirements
+		// actions are not checked since they are not necessarily done by the read module
+		readSelector = &modules.Selector{Flow: app.Read,
+			Destination:  sink,
+			Actions:      make([]pb.EnforcementAction, 0),
+			Source:       nil,
+			Dependencies: make([]*app.M4DModule, 0),
+			Module:       nil,
+			Message:      ""}
+		if !readSelector.SelectModule(m.Modules) {
+			m.Log.Info(item.AssetID + " : " + readSelector.GetError())
+			return instances, errors.New(readSelector.GetError())
+		}
+		// logic for deciding whether copy module is required
+		copyRequired, interfaces, actionsOnCopy = m.getCopyRequirements(item, readSelector)
+		// is copy allowed?
+		if !item.Actions[pb.AccessOperation_WRITE].Allowed {
+			return instances, errors.New(app.CopyNotAllowed)
+		}
+	} else {
+		// COPY FLOW
```

Comment:
@shlomitk1 Any time we're not doing READ we're assuming it's an ingest flow? 
I don't think anyone reading this code could figure out that we're loading new data into the environment.  "Copy flow" comment is not sufficient to understand this.

Reply:
OK. I'll return the flow argument to the structure.